Habitica/src/main/java/com/habitrpg/android/habitica/data/implementation/ChallengeRepositoryImpl.java

```diff
@@ -0,0 +1,123 @@
+            List<String> taskIdList = $.map(entry.getValue(), t -> t.getId());
+
+            switch (entry.getKey()) {
+                case Task.TYPE_HABIT:
+                    tasksOrder.setHabits(taskIdList);
+                    break;
+                case Task.TYPE_DAILY:
+                    tasksOrder.setDailys(taskIdList);
+                    break;
+                case Task.TYPE_TODO:
+                    tasksOrder.setTodos(taskIdList);
+                    break;
+                case Task.TYPE_REWARD:
+                    tasksOrder.setRewards(taskIdList);
+                    break;
+            }
+        }
+
+        return tasksOrder;
+    }
+
+    private Observable addChallengeTasks(String challengeId, List<Task> addedTaskList) {
+        if (addedTaskList.size() == 1) {
+            return apiClient.createChallengeTask(challengeId, addedTaskList.get(0));
+        } else {
+            return apiClient.createChallengeTasks(challengeId, addedTaskList);
+        }
+    }
+
+    @Override
+    public Observable<Challenge> createChallenge(PostChallenge challenge, List<Task> taskList) {
+        challenge.tasksOrder = getTaskOrders(taskList);
+
+        return Observable.create(subscriber -> {
+            apiClient.createChallenge(challenge).subscribe(challenge1 -> {
+                addChallengeTasks(challenge1.id, taskList).subscribe(task -> {
+                    subscriber.onNext(challenge1);
+                    subscriber.onCompleted();
+                }, throwable -> subscriber.onError((Throwable) throwable));
+
+            }, throwable -> subscriber.onError(throwable));
+        });
+    }
+
+    @Override
+    public Observable<Challenge> updateChallenge(PostChallenge challenge, List<Task> fullTaskList,
+                                                 List<Task> addedTaskList, List<Task> updatedTaskList, List<String> removedTaskList) {
+
+        ArrayList<Observable> observablesToWait = new ArrayList<>($.map(updatedTaskList, t -> apiClient.updateTask(t.getId(), t)));
+        observablesToWait.addAll($.map(removedTaskList, apiClient::deleteTask));
+
+        if (addedTaskList.size() != 0) {
+            observablesToWait.add(addChallengeTasks(challenge.id, addedTaskList));
+        }
+
+        challenge.tasksOrder = getTaskOrders(fullTaskList);
+
+
+        return Observable.create(subscriber -> {
+
+            Observable.from(observablesToWait)
+                    .flatMap(task -> task.subscribeOn(Schedulers.computation()))
+                    .toList()
+                    .subscribe(o -> {
+                        apiClient.updateChallenge(challenge)
+                                .subscribe(challenge1 -> {
+
+                                    subscriber.onNext(challenge1);
+                                    subscriber.onCompleted();
+                                }, throwable -> subscriber.onError(throwable));
+                    });
+        });
```

Comment:
thats what I had in first place, it has to look like this, since the `observableToWait` is just an array and `Observables.from()` not gonna wait for it, even with a `toList()`, you have to use something like this:

```java
return Observable.from(observablesToWait)
                .flatMap(task -> task.subscribeOn(Schedulers.computation()))
                .toList()
                .flatMap(tasks -> apiClient.updateChallenge(challenge));
```

but then I get this warning in the editor: 

`Unchecked assignment: 'rx.Observable' to 'rx.Observable<com.habitrpg.android.habitica.models.social.Challenge>'. Reason: 'Observable.from(observablesToWait) .flatMap(task -> task.subscribeOn(Schedulers.computation())) .toList()' has raw type, so result of flatMap is erased `


and now that I'm retrying to get my error I had, it works without any crashes 😕  

gonna use the statement up there